### DISCUSSION TITLE
[RFC] Refs naming standard

### DISCUSSION THREAD

**Discussion by zombieJ** [ant-design/ant-design#45242] [RFCs]
# 📋 RFC

## Summary

Align the `ref` API naming standard. All `ref` should provider `nativeElement` as dom of the root.

## Basic example

```tsx
const ref = useRef();

useEffect(() => {
  ref.current.scrollTo({ key: 'bamboo' });
  console.log(ref.current.nativeElement);
}, []);

<Table ref={ref} />
```

## Motivation

Current `ref` in antd is miss with operation and dom element. It's not extendable if already used for dom or operation.

## Detailed

Use `Proxy` to handle the `ref` logic for exist component but new object for not support `ref` component yet.

## Drawbacks

Not affect for current usage.

## Adoption strategy

Apply in minor version.

**Comment by babur001** [comment]
Any other viable use cases for accessing table's ref?

If it is just scroll or get components heigh or position, i can use just use simple `div` wrapper.
```
const ref = useRef<HtmlDivElement>();

useEffect(() => {
  ref.current.scrollTo({ key: 'bamboo' });
  console.log(ref.current.nativeElement);
}, []);

<div ref={ref}> 
   <Table />
</div> 
```

**Reply by zombieJ** [reply]
`key` should work with component logic. So you call in native element will not work. For Table, we've follow the RFC to extends with `scrollTo` function:
https://github.com/ant-design/ant-design/pull/45245

**Comment by gregor-mueller** [comment]
> Apply in minor version.

This broke a _lot_ of apps for us (providing a UI library based on antd). We use `ref` internally in `our-lib` and antd is a peer dependency, so consumers can have newer (minor and patch!) versions. We got a bit lucky, using the library ourselves and hitting the breaking change within our own app. So the UI lib has the following _package.json_:

```json
{
  "name": "our-lib",
  "peerDependencies": {
    "antd": "^5.10.0"
  },
  "dependencies": {
    "antd": "^5.10.0"
  }
}
``` 


The app using our library has the  following _package.json_:

```json
{
  "name": "some-app",
  "dependencies": {
    "our-lib": "latest"
  }
}
``` 

When doing a simple `npm i` it got served `antd@5.11.0` so the app crashed when trying to render the table with no easy way for them finding out, what was wrong.

We fixed this now, updating `our-lib` but the error occurred for others without us noticing.  
**Changes like this should always be in major versions only, IMHO.**
